### PR TITLE
[Issue #1491]: Setup states to hide clear all and select all in filter accordion

### LINKED ISSUE #1491: [Task]: Disable select all when all selected. Disable clear all when none selected.
### Summary

Disable select all when all selected. Disable clear all when none selected.  Can likely set the `disabled` prop (or attribute) here.

### Acceptance criteria

- You can't `Select All` twice in succession
- You can't `Clear All` twice in succession

### PR DESCRIPTION
**Pull request by rylew1**
## Summary
Fixes #1491 

## Changes proposed
- add states for tracking clear all / select all
    - states are passed down through accordion filter 
- move `searchfetcher` classes to `/services/search/searchfetcher` folder

## Context for reviewers

- I think we need to refactor `useSearchFilter` as it's getting a little hard to follow - created another ticket #1498 

##  Demo

https://github.com/HHS/simpler-grants-gov/assets/93001277/fa64071a-a44e-44b7-9d1f-55a10d9e324d

### CODE REVIEW

File: frontend/src/hooks/useSearchFilter.ts [hunk 1]
```diff
@@ -68,7 +149,54 @@ function useSearchFilter(initialFilterOptions: FilterOption[]) {
...
+          ...prevState,
+          [sectionId]: newSelectionStates.isSectionNoneSelected[sectionId],
+        }));
+      } else {
+        setIsSectionAllSelected(newSelectionStates.isSectionAllSelected);
+        setIsSectionNoneSelected(newSelectionStates.isSectionNoneSelected);
+      }
+    },
+    [determineInitialSelectionStates],
+  );
```
Comment (rylew1) [inline frontend/src/hooks/useSearchFilter.ts:197]: This is the new code in just this PR

File: frontend/src/app/search/page.tsx [hunk 2]
```diff
@@ -10,7 +10,7 @@ import SearchCallToAction from "../../components/search/SearchCallToAction";
 import { SearchForm } from "./SearchForm";
 import { convertSearchParamsToProperTypes } from "../../utils/convertSearchParamsToStrings";
 import { cookies } from "next/headers";
-import { getSearchFetcher } from "../../services/searchfetcher/SearchFetcherUtil";
+import { getSearchFetcher } from "../../services/search/searchfetcher/SearchFetcherUtil";
```
Comment (rylew1) [inline frontend/src/app/search/page.tsx:13]: I also moved the `SearchFetcher` classes from `/services/searchfetcher` to `services/search/searchfetcher/` so there's a bunch of updates in this PR for that

File: frontend/src/services/search/SearchFilterManager.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      isChecked: initialParamsSet.has(option.value),
+      children: option.children
+        ? option.children.map((child) => ({
+            ...child,
+            isChecked: initialParamsSet.has(child.value),
+          }))
+        : undefined,
+    }));
+  };
+}
```
Comment (rylew1) [inline frontend/src/services/search/SearchFilterManager.ts:46]: Start the creation of this class (currently not used but I think we should start to move some of the management of checked options here (how many total checked, how many in a section are checked, if all in a section are checked are not) - to make it easy to just grab these values instead of looping through all the options everytime. 

It probably won't lead to a ton of performance improvement though - just more for our own sanity :)

File: frontend/src/hooks/useSearchFilter.ts [hunk 4]
```diff
@@ -43,6 +44,69 @@ function useSearchFilter(
...
+  const [isNoneSelected, setIsNoneSelected] = useState<boolean>(
+    initialSelectionStates.noneSelected,
+  );
+  const [isSectionAllSelected, setIsSectionAllSelected] = useState<{
+    [key: string]: boolean;
+  }>(initialSelectionStates.isSectionAllSelected);
+  const [isSectionNoneSelected, setIsSectionNoneSelected] = useState<{
+    [key: string]: boolean;
+  }>(initialSelectionStates.isSectionNoneSelected);
+
```
Comment (rylew1) [inline frontend/src/hooks/useSearchFilter.ts:109]: loads of logic to manage the {select,clear} all disabled state

File: frontend/src/services/search/SearchFilterManager.ts (hunk 3 above)
Comment (rylew1) [reply frontend/src/services/search/SearchFilterManager.ts:46]: Might try this in a part 2 PR on this #1491 ticket

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx [hunk 5]
```diff
@@ -84,6 +90,8 @@ export function SearchFilterAccordion({
                 mounted={mounted}
                 updateCheckedOption={toggleOptionChecked}
                 toggleSelectAll={toggleSelectAll}
+                isSectionAllSelected={isSectionAllSelected[option.id]}
+                isSectionNoneSelected={isSectionNoneSelected[option.id]}
               />
```
Comment (andycochran) [inline frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx:95]: Nice.

File: frontend/src/services/search/SearchFilterManager.ts (hunk 3 above)
Comment (andycochran) [reply frontend/src/services/search/SearchFilterManager.ts:46]: Cool. Should we put a TODO in this file noting as much (that it's not being used yet and a link to a ticket to implement/explore)?

**Review by andycochran — APPROVED** [review]

File: frontend/src/services/search/SearchFilterManager.ts (hunk 3 above)
Comment (rylew1) [reply frontend/src/services/search/SearchFilterManager.ts:46]: yea good catch - will add that!

File: frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx (hunk 5 above)
Comment (rylew1) [reply frontend/src/components/search/SearchFilterAccordion/SearchFilterAccordion.tsx:95]: Yea the logic to setup those values is kind of a lot but once you have it it's very easy to use in the components
